### DISCUSSION TITLE
System.Text.Json missing streaming reader in runtime / adopt existing implementation?

### DISCUSSION THREAD

**Discussion by bjornharrtell** [dotnet/runtime#100439] [Ideas]
Hi,

I've since some time implemented streaming readers based on `System.Text.Json` that has been in production use for use cases with very large JSON payloads retrieved via stream. Source lives at https://github.com/bjornharrtell/Utf8JsonStreamReader and published at nuget at https://www.nuget.org/packages/Utf8JsonStreamReader and ~~benchmarks has shown it's faster than corresponding NewtonsoftJson based logic~~ UPDATE: recent experiments suggest sync variant of Newtonsoft streaming reader is actually faster.

I recall being surprised something similar wasn't already in the runtime. Actually I find it strangely asymmetric that the official `Utf8JsonWriter` is a class that can write to stream but `Utf8JsonReader` is a ref struct and can only work with in memory buffer.

I wonder if there are interest in officially adopting my implementation?

**Comment by huoyaoyuan** [comment]
Being async at every read/write operation is not performant.
`Utf8JsonWriter` is synchronous at every write. It writes to its internal buffer and flushes asynchronously when you call `FlushAsync`.
`Utf8JsonReader` is synchronous at every read. However there's not public utility that helps you reload buffer asynchronously when you reach the end.

Such utility should expose synchronous `TryRead` and asynchronously `FillBufferAsync`.

There's also space for improvement in your shape of implementation:
- `ReadAsync` should be `ValueTask` based. Since there's a high chance of completing synchronously, using `ValueTask` will provide significant benefit even when not pooling manually.
- Returning boxed value by default will increase boxing significantly.
- It's not clear to me how your implementation handles huge string tokens that is larger than your buffer size.

**Reply by bjornharrtell** [reply]
Thanks for the feedback @huoyaoyuan. I agree with your observations and will explore possible improvements when I can.

**Reply by bjornharrtell** [reply]
@huoyaoyuan the Utf8JsonStreamReader class has a sync Read and so does Newtonsoft. I added benchmarks using those and the results verify that it can perform better. In the case of Newtonsoft *much* better and beating my implementation by a large margin, which I did not expect... 


```sh
BenchmarkDotNet v0.13.12, Ubuntu 23.10 (Mantic Minotaur)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.103
  [Host]     : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method                                | Objects | Mean     | Error   | StdDev  |
|-------------------------------------- |-------- |---------:|--------:|--------:|
| TraverseUtf8JsonStreamReader          | 100000  | 171.9 ms | 0.58 ms | 0.49 ms |
| TraverseUtf8JsonStreamReaderSync      | 100000  | 161.2 ms | 1.52 ms | 1.27 ms |
| TraverseUtf8JsonStreamTokenEnumerator | 100000  | 147.3 ms | 2.88 ms | 5.33 ms |
| TraverseNewtonsoftJsonTextReader      | 100000  | 182.4 ms | 0.96 ms | 0.85 ms |
| TraverseNewtonsoftJsonTextReaderSync  | 100000  | 109.8 ms | 0.27 ms | 0.25 ms |
```

**Reply by bjornharrtell** [reply]
With some rewrites and improvements I have been able to at least again beat Newtonsoft:

```sh
BenchmarkDotNet v0.13.12, Ubuntu 23.10 (Mantic Minotaur)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.103
  [Host]     : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method                                     | Objects | Mean     | Error   | StdDev  |
|------------------------------------------- |-------- |---------:|--------:|--------:|
| TraverseUtf8JsonStreamReaderAsync          | 100000  | 138.9 ms | 0.61 ms | 0.48 ms |
| TraverseUtf8JsonStreamReader               | 100000  | 101.8 ms | 0.36 ms | 0.30 ms |
| TraverseUtf8JsonStreamTokenAsyncEnumerable | 100000  | 139.0 ms | 2.66 ms | 3.56 ms |
| TraverseUtf8JsonStreamTokenEnumerable      | 100000  | 100.1 ms | 0.45 ms | 0.40 ms |
| TraverseNewtonsoftJsonTextReaderAsync      | 100000  | 185.4 ms | 0.75 ms | 0.62 ms |
| TraverseNewtonsoftJsonTextReader           | 100000  | 110.9 ms | 0.62 ms | 0.58 ms |
```

**Reply by bjornharrtell** [reply]
On "Returning boxed value by default will increase boxing significantly" can you explain some more? Do you mean the fact that I decode value up front fx. string? That is costly I agree and a waste especially if the user is only using partial content, but I've not figured out how to make a nice interface to leave it up to caller to decode due to the ref struct nature of Utf8JsonReader .

About huge tokens I do not make any attempt to handle it, it will fail and user will have to supply a larger buffer size.

**Reply by bjornharrtell** [reply]
I've realised the my design/implementation was flawed and buffering up results isn't effective, at least not in terms of memory footprint. I've reworked into delegate interface now which seems to be both faster and small memory footprint.

**Reply by bjornharrtell** [reply]
Latest benchmark:

```sh
BenchmarkDotNet v0.13.12, Ubuntu 23.10 (Mantic Minotaur)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.103
  [Host]     : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method                               | Objects | Mean      | Error    | StdDev   |
|------------------------------------- |-------- |----------:|---------:|---------:|
| TraverseUtf8JsonStreamReader         | 100000  |  62.55 ms | 0.676 ms | 0.632 ms |
| TraverseUtf8JsonStreamReaderRawValue | 100000  |  28.15 ms | 0.139 ms | 0.124 ms |
| TraverseNewtonsoftJsonTextReader     | 100000  | 108.26 ms | 0.840 ms | 0.786 ms |
```

**Reply by huoyaoyuan** [reply]
@eiriktsarpalis With STJ getting Pipelines support, is there anything applicable to this?

**Reply by davidfowl** [reply]
This might help https://devblogs.microsoft.com/dotnet/the-convenience-of-system-text-json/

**Reply by eiriktsarpalis** [reply]
> With STJ getting Pipelines support, is there anything applicable to this?

The pipelines APIs will be added to `JsonSerializer`, which lives higher in the stack. Presumably this question concerns the low-level reader type.
